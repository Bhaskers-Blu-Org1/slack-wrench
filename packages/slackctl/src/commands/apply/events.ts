import { WebClient } from '@slack/web-api';
import url from 'url';

import { Config } from '../../config';

export default async function apply(config: Config, slack: WebClient) {
  const { id: app, events, baseUrl } = config;

  if (!events) {
    return;
  }

  const eventsCallback = url.resolve(baseUrl, events.requestPath);

  try {
    await slack.apiCall('developer.apps.events.subscriptions.verifyURL', {
      app,
      url: eventsCallback,
    });
  } catch (error) {
    throw new Error(
      `Couldn't verify event url ${eventsCallback}. Do you have a server running?`,
    );
  }

  await slack.apiCall('developer.apps.events.subscriptions.updateSubs', {
    app,
    app_event_types: events.workspaceEvents || [],
    bot_event_types: events.botEvents || [],
    enable: true,
    filter_teams: [],
    unfurl_domains: events.unfurlDomains || [],
    url: eventsCallback,
  });

  console.log('Receiving Slack events at: ', eventsCallback);
}
