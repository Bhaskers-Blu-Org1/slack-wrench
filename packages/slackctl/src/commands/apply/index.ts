import path from 'path';
import { CommandBuilder, Argv } from 'yargs';

import { Config } from '../../config';
import loadConfig from '../../load-config';

interface ApplyArgs {
  config: string;
  timeout: number;
  ngrok?: number;
}

const handler = (argv: ApplyArgs): Promise<Config> => {
  const searchFrom = path.resolve(process.cwd(), argv.config);
  return loadConfig(searchFrom);
};

const builder = (command: any): any =>
  command
    .option('n', {
      alias: 'ngrok',
      type: 'number',
    })
    .option('t', {
      alias: 'timeout',
      type: 'number',
      default: 0,
    })
    .option('c', {
      alias: 'config',
      type: 'string',
      default: process.cwd(),
    }) as CommandBuilder<any, ApplyArgs>;

const command = {
  command: 'apply',
  describe: 'apply a Slack app configuration',
  builder,
  handler,
};

export default command;
