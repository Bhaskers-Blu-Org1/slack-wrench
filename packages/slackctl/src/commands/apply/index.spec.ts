import path from 'path';
import yargs from 'yargs';

import apply from './index';

describe('The apply command', () => {
  const parser = yargs.command(apply).help();

  const run = (cmd: string) =>
    new Promise(resolve => {
      parser.parse(cmd, (err: Error, argv: any, output: string) => {
        resolve(output);
      });
    });

  it('should output help', async () => {
    expect.assertions(1);
    const output = await run('apply --help');

    expect(output).toMatchSnapshot();
  });

  it('should load a correct config', async () => {
    expect.assertions(1);
    const config = await apply.handler({
      config: path.resolve(__dirname, '../../test/fixtures/correct'),
      timeout: 0,
    });

    expect(config).toMatchSnapshot();
  });
});
