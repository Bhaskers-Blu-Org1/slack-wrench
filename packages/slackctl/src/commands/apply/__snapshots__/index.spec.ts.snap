// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The apply command should load a correct config 1`] = `
Object {
  "baseUrl": "https://example.com",
  "events": Object {
    "botEvents": Array [
      "app_mention",
      "message.channels",
      "message.groups",
    ],
    "requestPath": "/slack/events",
    "unfurlDomains": Array [
      "ibm.com",
      "twitter.com",
    ],
    "workspaceEvents": Array [
      "message.im",
    ],
  },
  "id": "AKH8DK7J7",
  "sessionToken": "xoxs-8709467270-220587980003-938427407255-f3fa6bd49e60c1516838b8dea08ade749e2f7869301020141695f6db629749fd",
  "version": "v1.0",
}
`;

exports[`The apply command should output help 1`] = `
"jest apply

apply a Slack app configuration

Options:
  --version      Show version number                                   [boolean]
  --help         Show help                                             [boolean]
  -n, --ngrok                                                           [number]
  -t, --timeout                                            [number] [default: 0]
  -c, --config                                                [string] [default:
                  \\"/Users/michael/workspace/ibm/slack-wrench/packages/slackctl\\"]"
`;
